File: xtee6/center-common/app/models/global_group.rb
#
# The MIT License
# Copyright (c) 2015 Estonian Information System Authority (RIA), Population Register Centre (VRK)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

class GlobalGroup < ActiveRecord::Base
  validates_with Validators::MaxlengthValidator
  validates :group_code, :presence => true, :uniqueness => true

  has_many :global_group_members, :dependent => :destroy

  # Adds a member to the group. The member should not be saved
  # to database (or loaded from database)
  def add_member(client_id)
    # Check if global group already contains the ID.
    if !has_member?(client_id)
      global_group_members.create!(:group_member => client_id.clean_copy())
    end

    update_member_count()
  end

  # Removes member from the group
  def remove_member(client_id)
    removable_members = GlobalGroupMember.joins(:group_member).where(
        :global_group_id => id,
        :identifiers => client_id_parameters(client_id))

    for to_remove in removable_members
      global_group_members.destroy(to_remove)
    end

    update_member_count()
  end

  # Returns true, if group has member with given ID.
  def has_member?(client_id)
    GlobalGroupMember.joins(:group_member).exists?(
        :global_group_id => id,
        :identifiers => client_id_parameters(client_id))
  end

  # Returns global group that contains security server owners.
  # Returns null, if system parameter is not defined or the
  # group does not exist.
  def self.security_server_owners_group
    group_code = SystemParameter.security_server_owners_group
    if group_code == nil
      nil
    else
      first(:conditions => {:group_code => group_code})
    end
  end

  def self.find_by_code(group_code)
     groups = GlobalGroup.where(:group_code => group_code)
     groups.empty? ? nil : groups[0]
  end

  def self.get_groups(query_params)
    get_search_relation(query_params.search_string).
        order("#{query_params.sort_column} #{query_params.sort_direction}").
        limit(query_params.display_length).
        offset(query_params.display_start)
  end

  def self.get_group_count(searchable)
    searchable.empty? ?
        GlobalGroup.count :
        get_search_relation(searchable).count
  end

  def self.add_group(code, description)
    GlobalGroup.create!(:group_code => code, :description => description)
  end

  def self.add_group_if_not_exists(code, description)
    find_by_code(code) == nil ? add_group(code, description) : nil
  end

  def self.update_description(group_id, description)
    group = GlobalGroup.find(group_id)
    group.description = description
    group.save!
  end

  def self.get_member_count(group_id)
    GlobalGroupMember.where(:global_group_id => group_id).count
  end

  private

  def client_id_parameters(client_id)
    { :object_type => client_id.object_type,
      :xroad_instance => client_id.xroad_instance,
      :member_class => client_id.member_class,
      :member_code => client_id.member_code,
      :subsystem_code => client_id.subsystem_code }
  end

  def update_member_count
    update_attributes!(:member_count => global_group_members.size)
  end

  def self.get_search_relation(searchable)
    sql_generator =
        SimpleSearchSqlGenerator.new(get_searchable_columns(), searchable)

    GlobalGroup.where(sql_generator.sql, *sql_generator.params)
  end

  def self.get_searchable_columns
    return [
        "global_groups.group_code",
        "global_groups.description",
        "global_groups.updated_at"]
  end
end

File: src/centralserver/admin-service/src/main/java/org/niis/xroad/centralserver/restapi/entity/GlobalGroup.java
/**
 * The MIT License
 *
 * Copyright (c) 2019- Nordic Institute for Interoperability Solutions (NIIS)
 * Copyright (c) 2018 Estonian Information System Authority (RIA),
 * Nordic Institute for Interoperability Solutions (NIIS), Population Register Centre (VRK)
 * Copyright (c) 2015-2017 Estonian Information System Authority (RIA), Population Register Centre (VRK)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.niis.xroad.centralserver.restapi.entity;

import lombok.Getter;
import lombok.Setter;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Transient;

import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = GlobalGroup.TABLE_NAME)
public class GlobalGroup extends AuditableEntity {

    public static final String TABLE_NAME = "global_groups";

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = TABLE_NAME + "_id_seq")
    @SequenceGenerator(name = TABLE_NAME + "_id_seq", sequenceName = TABLE_NAME + "_id_seq", allocationSize = 1)
    @Column(name = "id", unique = true, nullable = false)
    @Getter
    private int id;

    @Column(name = "group_code")
    @Getter
    @Setter
    private String groupCode;

    @Column(name = "description")
    @Getter
    @Setter
    private String description;

    @Column(name = "member_count")
    @Access(AccessType.FIELD)
    private Integer memberCount;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "globalGroup", cascade = CascadeType.ALL, orphanRemoval = true)
    @Access(AccessType.FIELD)
    @Getter
    private Set<GlobalGroupMember> globalGroupMembers = new HashSet<>(0);

    protected GlobalGroup() {
        //JPA
    }

    public GlobalGroup(String groupCode) {
        this.groupCode = groupCode;
    }

    @Transient
    public int getMemberCount() {
        return this.memberCount == null ? 0 : this.memberCount;
    }
}
